core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java

```diff
@@ -1507,69 +1507,58 @@ private SqlNode getCastedSqlNode(SqlNode argInput, RelDataType varType,
     }
   }
 
-  /** Convertlet that handles Oracle's custom {@code Substr} function. */
-  private static class OracleSubStrConvertlet implements SqlRexConvertlet {
+  /** Convertlet that handles Oracle's {@code SUBSTR} function. */
+  private static class OracleSubstrConvertlet implements SqlRexConvertlet {
     @Override public RexNode convertCall(SqlRexContext cx, SqlCall call) {
       // Translate
-      //   substr(value, start, length)
+      //   SUBSTR(value, start, length)
       // to
-      //   CASE
-      //   WHEN length <= 0
-      //   THEN NULL
-      //   ELSE substring(
-      //     value, (
-      //       CASE
-      //       WHEN index = 0
-      //       THEN 1
-      //       WHEN index < 0
-      //       THEN length(value) - start + 1
-      //       ELSE start),
-      //     length
-      //   END
+      //   SUBSTRING(
```

Comment:
Substring behavior with negative indices, seem ill defined. For instances, postgres does:

```
postgres=# SELECT SUBSTRING('abc', -1, 3);
substring
a
(1 row)
```

So, I think the out of bound length check should be handled in an outer case statement instead of rewriting the start index
CASE WHEN length(value) + start + 1 > 0 THEN '' .... )

Reply:
The behavior of the SQL standard `SUBSTRING` function is well-defined. Literally. Because it's defined in the standard. Calcite's SUBSTRING function is based on that, not MySQL's or PostgreSQL's `SUBSTR` (or Java's `substring` for that matter).

Given that, I was trying to produce the simplest Rex expression that could work (i.e. pass all tests).

Reply:
According to SQL199 page 169, the start index is greater of  specified start position and 1.  This is clearly not what calcite has implemented.  

My suggested change would not be substantively more complicated than the existing code.  And would correctly handle translating oracle queries too postgres.